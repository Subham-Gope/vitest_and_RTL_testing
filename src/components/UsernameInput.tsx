import React, { useState } from "react";
import type { ChangeEvent, FocusEvent } from "react";

const UsernameInput: React.FC = () => {
  const [focused, setFocused] = useState<boolean>(false);
  const [value, setValue] = useState<string>("");

  const handleFocus = (e: FocusEvent<HTMLInputElement>): void => {
    setFocused(true);
    console.log(e);
  };
  const handleBlur = (e: FocusEvent<HTMLInputElement>): void => {
    setFocused(false);
    console.log(e);
  };

  const handleChange = (e: ChangeEvent<HTMLInputElement>) => {
    setValue(e.target.value);
  };

  return (
    <>
      <div>
        <label htmlFor="focus-blur">Enter your username:</label>
        <input
          type="text"
          placeholder="username"
          id="focus-blur"
          value={value}
          onChange={handleChange}
          onFocus={handleFocus}
          onBlur={handleBlur}
        />
      </div>
      {focused && <p>Username must be atleast 4 characters</p>}
    </>
  );
};

export default UsernameInput;
