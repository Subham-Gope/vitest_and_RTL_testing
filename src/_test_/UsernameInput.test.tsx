import { describe, test, expect } from "vitest";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import UsernameInput from "../components/UsernameInput";

describe("Testing the UsernameInput.tsx file", () => {
  test("focuses on a element and then blurs out", async () => {
    // setting up the userEvent
    const user = userEvent.setup();

    // setting up the debugger
    screen.debug();

    // setting up the snapshot matching
    const { container } = render(<UsernameInput />);
    expect(container).toMatchSnapshot();

    const input = screen.getByRole("textbox", {
      name: /enter your username:/i,
    });
    expect(input).toBeInTheDocument();

    expect(input).not.toHaveFocus();

    // Initiating the user interactions
    await user.click(input);
    expect(input).toHaveFocus();

    const displayText = screen.getByText(/username must be/i);
    expect(displayText).toBeInTheDocument();

    await user.tab();
    expect(input).not.toHaveFocus();
    expect(displayText).not.toBeInTheDocument();
  });
});
