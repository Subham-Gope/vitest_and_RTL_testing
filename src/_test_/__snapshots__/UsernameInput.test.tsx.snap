// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Testing the UsernameInput.tsx file > focuses on a element and then blurs out 1`] = `
<div>
  <div>
    <label
      for="focus-blur"
    >
      Enter your username:
    </label>
    <input
      id="focus-blur"
      placeholder="username"
      type="text"
      value=""
    />
  </div>
</div>
`;
